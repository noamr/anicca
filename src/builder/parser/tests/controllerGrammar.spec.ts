/// </ reference "jest" />
import nearley from 'nearley'
const controllerGrammar = require('../controller.ne')

const parse = (str: string) => new nearley.Parser(controllerGrammar).feed(str).finish()[0]

describe('controller grammar', () => {
    it('state', () => {
        expect(parse('state myState')).toMatchSnapshot()
        expect(parse('parallel myState')).toMatchSnapshot()
        expect(parse('history myState')).toMatchSnapshot()
        expect(parse('deep history myState')).toMatchSnapshot()
        expect(parse('shallow history myState')).toMatchSnapshot()
        expect(parse('final myState')).toMatchSnapshot()
        expect(parse('initial')).toMatchSnapshot()
    })

    it('transition', () => {
        expect(parse('on someEvent')).toMatchSnapshot()
        expect(parse('when someCondition + (a - 1) * grumpy()')).toMatchSnapshot()
        expect(parse('on someEvent when a == 1')).toMatchSnapshot()
        expect(parse('when true')).toMatchSnapshot()
        expect(parse('after 10 s when a == 1')).toMatchSnapshot()
        expect(parse('after 30 minutes')).toMatchSnapshot()
    })
    it('payload', () => {
        expect(parse('on completed(payload as u32)')).toMatchSnapshot()
        expect(parse('on completed(a as string, b as bool)')).toMatchSnapshot()
        expect(parse('on done()')).toMatchSnapshot()
        expect(parse('on start(payload as SomeType) when a == 1')).toMatchSnapshot()
    })
    it('upon', () => {
        expect(parse('entering')).toMatchSnapshot()
        expect(parse('leaving')).toMatchSnapshot()
    })
})
