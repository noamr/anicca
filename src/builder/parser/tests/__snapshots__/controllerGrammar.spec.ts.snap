// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`controller grammar payload 1`] = `
Object {
  "event": "completed",
  "payload": Object {
    "payload": Array [
      0,
      "u32",
    ],
  },
  "type": "Transition",
}
`;

exports[`controller grammar payload 2`] = `
Object {
  "event": "completed",
  "payload": Object {
    "a": Array [
      0,
      "string",
    ],
    "b": Array [
      1,
      "bool",
    ],
  },
  "type": "Transition",
}
`;

exports[`controller grammar payload 3`] = `
Object {
  "event": "done",
  "payload": Object {},
  "type": "Transition",
}
`;

exports[`controller grammar payload 4`] = `
Object {
  "condition": Object {
    "args": Array [
      Object {
        "$ref": "a",
      },
      Object {
        "$primitive": 1,
      },
    ],
    "op": "eq",
  },
  "event": "start",
  "payload": Object {
    "payload": Array [
      0,
      "SomeType",
    ],
  },
  "type": "Transition",
}
`;

exports[`controller grammar state 1`] = `
Object {
  "name": "myState",
  "type": "State",
}
`;

exports[`controller grammar state 2`] = `
Object {
  "name": "myState",
  "type": "Parallel",
}
`;

exports[`controller grammar state 3`] = `
Object {
  "deep": true,
  "name": "myState",
  "type": "History",
}
`;

exports[`controller grammar state 4`] = `
Object {
  "deep": true,
  "name": "myState",
  "type": "History",
}
`;

exports[`controller grammar state 5`] = `
Object {
  "deep": false,
  "name": "myState",
  "type": "History",
}
`;

exports[`controller grammar state 6`] = `
Object {
  "name": "myState",
  "type": "Final",
}
`;

exports[`controller grammar state 7`] = `
Object {
  "type": "Initial",
}
`;

exports[`controller grammar transition 1`] = `
Object {
  "event": "someEvent",
  "type": "Transition",
}
`;

exports[`controller grammar transition 2`] = `
Object {
  "condition": Object {
    "args": Array [
      Object {
        "$ref": "someCondition",
      },
      Object {
        "args": Array [
          Object {
            "args": Array [
              Object {
                "$ref": "a",
              },
              Object {
                "$primitive": 1,
              },
            ],
            "op": "sub",
          },
          Object {
            "args": Array [],
            "op": "grumpy",
          },
        ],
        "op": "mult",
      },
    ],
    "op": "add",
  },
  "type": "Transition",
}
`;

exports[`controller grammar transition 3`] = `
Object {
  "condition": Object {
    "args": Array [
      Object {
        "$ref": "a",
      },
      Object {
        "$primitive": 1,
      },
    ],
    "op": "eq",
  },
  "event": "someEvent",
  "type": "Transition",
}
`;

exports[`controller grammar transition 4`] = `
Object {
  "condition": Object {
    "$primitive": true,
  },
  "type": "Transition",
}
`;

exports[`controller grammar transition 5`] = `
Object {
  "condition": Object {
    "args": Array [
      Object {
        "$ref": "a",
      },
      Object {
        "$primitive": 1,
      },
    ],
    "op": "eq",
  },
  "timeout": Object {
    "args": Array [
      Object {
        "$primitive": 10,
      },
      Object {
        "$primitive": 1000,
      },
    ],
    "op": "mult",
  },
  "type": "Transition",
}
`;

exports[`controller grammar transition 6`] = `
Object {
  "timeout": Object {
    "args": Array [
      Object {
        "$primitive": 30,
      },
      Object {
        "$primitive": 60000,
      },
    ],
    "op": "mult",
  },
  "type": "Transition",
}
`;

exports[`controller grammar upon 1`] = `
Object {
  "type": "OnEntry",
}
`;

exports[`controller grammar upon 2`] = `
Object {
  "type": "OnExit",
}
`;
